### ISSUE TITLE
Gateway Code to start/stop relay

### ISSUE THREAD

**Issue by marcoherbst** [evercam/evercam-gateway#42]
We need some code on the Gateway, so it knows to start a particular
relay, i.e. starting ingesting with Gstreamer and relaying with Janus.
It needs to know which device to talk to. Also the Gateway code has to
be able to stop the relay too.

**Comment by scandox** [comment]
@rajmohanb @marcoherbst So this is the code that needs to run within the Gateway application itself. It is very important to understand that the Gateway is am umbrella App. That means that it has 2 apps inside it which share some of the same codebase. The Gateway App itself is in the sub-folder `apps/gateway`.

A few observations about the Gateway app:

*   It runs an `init` process before starting any of its functions. This either loads an existing configuration, or enters a registration process which doesn't end until the Gateway is approved.

*    The registration process only needs to complete once in the lifetime of a Gateway. Once that is complete, it will just load the config and start the application's multiple sub-processes.

*     The entry point for all the main activities of the Gateway (after config is loaded) is `lib/supervision/supervisor.ex`. It kicks off the various supervisors that monitor Gateway activities.

*     Main gateway activities are: Discovery and Routing. I envision the WebRTC functionality being two activities: Streaming and Relay. So we would have a folder of modules for controlling Gstreamer and a folder of modules for controlling Janus. Then we would have a single module called WebRTC that joins them together. This is only a proposal, but it would roughly match the current way of dividing things up.

*    @rajmohanb It may make sense to build this Elixir functionality outside of the Gateway and then we can merge it when it has a bit of a shape?

**Comment by rajmohanb** [comment]
@scandox I am OK with building it as a separate Elixir app and then integrate with existing gateway.

Also does it make sense to run this WebRTC functionality as another app within the umbrella app? However in case we decide to integrate within the gateway app, we could init the new functionality within `start_workers()` within `lib/supervision/supervisor.ex.`

**Comment by scandox** [comment]
@rajmohanb An important clarification: an umbrella app does NOT run its various apps in tandem. They are completely separate applications, that just happen to share some logic. If the logic was of a very general purpose then it would be made into a library and the two apps would be separate. In fact, the logic is quite specific to our case, so it is simply easier to have a single development bundle. But both these apps are totally separate.

So if you created a WebRTC app, then it would be designed to run separately from Gateway code. I think that isn't a good idea, because then we would have to find ways to share configs etc...

**Comment by rajmohanb** [comment]
@scandox OK, get the point. So we will run the WebRTC within the existing gateway app. So to start off, should this be built as a separate elixir app right now and then we will think of merging?

**Comment by scandox** [comment]
@rajmohanb yes. Build separately. It will keep your head clearer I think. It means you don't have to learn all about the Gateway first.

**Comment by rajmohanb** [comment]
Note: To support multiple cameras connected to the gateway, there is a need to support multiple simultaneous gstreamer processes.

**Comment by rajmohanb** [comment]
The separate elixir application is @ https://github.com/rajmohanb/webrtc_gw. I have verified that the basic demo works with this application for chrome.
